src/main/java/ch/njol/skript/expressions/ExprTool.java

```diff
@@ -42,75 +30,120 @@
+})
+@Since("1.0, 2.2-dev37 (offhand)")
 public class ExprTool extends PropertyExpression<LivingEntity, Slot> {
+
 	static {
 		Skript.registerExpression(ExprTool.class, Slot.class, ExpressionType.PROPERTY,
-			"[the] ((tool|held item|weapon)|1¦(off[ ]hand (tool|item))) [of %livingentities%]",
-			"%livingentities%'[s] ((tool|held item|weapon)|1¦(off[ ]hand (tool|item)))");
+				"[the] ((tool|held item|weapon)|offhand:(off[ ]hand (tool|item))) [of %livingentities%]",
+				"%livingentities%'[s] ((tool|held item|weapon)|1¦(off[ ]hand (tool|item)))"
+		);
 	}
 
 	private boolean offHand;
 
-	@SuppressWarnings({"unchecked", "null"})
 	@Override
-	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parser) {
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
 		setExpr((Expression<LivingEntity>) exprs[0]);
-		offHand = parser.mark == 1;
+		offHand = parseResult.hasTag("offhand");
 		return true;
 	}
 
 	@Override
-	protected Slot[] get(final Event e, final LivingEntity[] source) {
-		final boolean delayed = Delay.isDelayed(e);
+	protected Slot[] get(Event event, LivingEntity[] source) {
+		boolean delayed = Delay.isDelayed(event);
 		return get(source, new Getter<Slot, LivingEntity>() {
 			@Override
 			@Nullable
-			public Slot get(final LivingEntity ent) {
+			public Slot get(LivingEntity entity) {
 				if (!delayed) {
-					if (!offHand && e instanceof PlayerItemHeldEvent && ((PlayerItemHeldEvent) e).getPlayer() == ent) {
-						final PlayerInventory i = ((PlayerItemHeldEvent) e).getPlayer().getInventory();
-						return new InventorySlot(i, getTime() >= 0 ? ((PlayerItemHeldEvent) e).getNewSlot() : ((PlayerItemHeldEvent) e).getPreviousSlot());
-					} else if (e instanceof PlayerBucketEvent && ((PlayerBucketEvent) e).getPlayer() == ent) {
-						final PlayerInventory i = ((PlayerBucketEvent) e).getPlayer().getInventory();
-						boolean isOffHand = ((PlayerBucketEvent) e).getHand() == org.bukkit.inventory.EquipmentSlot.OFF_HAND || offHand;
-						return new InventorySlot(i, isOffHand ? EquipmentSlot.EquipSlot.OFF_HAND.slotNumber
-							: ((PlayerBucketEvent) e).getPlayer().getInventory().getHeldItemSlot()) {
+					if (offHand && event instanceof InventoryClickEvent && ((InventoryClickEvent) event).getWhoClicked().equals(entity) && getTime() == 1) {
+						InventoryClickEvent inventoryClickEvent = (InventoryClickEvent) event;
+						// When a player uses a number key to swap an item from hotbar to offhand. This simplifies the process with future states.
+						if (inventoryClickEvent.getClick() == ClickType.NUMBER_KEY && inventoryClickEvent.getSlot() == EquipmentSlot.EquipSlot.OFF_HAND.slotNumber) {
+							PlayerInventory inventory = inventoryClickEvent.getWhoClicked().getInventory();
+							return new InventorySlot(inventory, inventoryClickEvent.getHotbarButton());
```

Comment:
How would this ever be called? Number key click type is for 0 through 8, but off hand slot number is 40

Reply:
The event is when they press F (default keybind)

Javadoc:
```
The "swap item with offhand" key (defaults to F).
```

Skript builds a reference to slot 40 here yes. That's what will be needed for passing to Inventory.getItem(int)

Reply:
ohh i see, hovering over the offhand slot and pressing the number key for the selected tool. I misunderstood the getSlot method's return value, my bad.

I don't think this has anything to do with the f keybind though. That would be a different ClickType.